provider/provider_config/spark_config.go

```diff
@@ -0,0 +1,217 @@
+	"github.com/mitchellh/mapstructure"
+)
+
+type SparkExecutorType string
+
+const (
+	EMR        SparkExecutorType = "EMR"
+	Databricks SparkExecutorType = "DATABRICKS"
+)
+
+type AWSCredentials struct {
+	AWSAccessKeyId string
+	AWSSecretKey   string
+}
+
+type SparkExecutorConfig interface {
+	Serialize() ([]byte, error)
+	Deserialize(config SerializedConfig) error
+	IsExecutorConfig() bool
+}
+
+type SparkFileStoreConfig interface {
+	Serialize() ([]byte, error)
+	Deserialize(config SerializedConfig) error
+	IsFileStoreConfig() bool
+}
+
+type SparkConfig struct {
+	ExecutorType   SparkExecutorType
+	ExecutorConfig SparkExecutorConfig
+	StoreType      FileStoreType
+	StoreConfig    SparkFileStoreConfig
+}
+
+func (s *SparkConfig) Deserialize(config SerializedConfig) error {
+	err := json.Unmarshal(config, s)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (s *SparkConfig) Serialize() ([]byte, error) {
+	conf, err := json.Marshal(s)
+	if err != nil {
+		return nil, err
+	}
+	return conf, nil
+}
+
+func (s *SparkConfig) UnmarshalJSON(data []byte) error {
+	type tempConfig struct {
+		ExecutorType   SparkExecutorType
+		ExecutorConfig map[string]interface{}
+		StoreType      FileStoreType
+		StoreConfig    map[string]interface{}
+	}
+
+	var temp tempConfig
+	err := json.Unmarshal(data, &temp)
+	if err != nil {
+		return fmt.Errorf("unmarshal: %w", err)
+	}
+
+	s.ExecutorType = temp.ExecutorType
+	s.StoreType = temp.StoreType
+
+	err = s.decodeExecutor(temp.ExecutorType, temp.ExecutorConfig)
+	if err != nil {
+		return fmt.Errorf("could not decode executor: %w", err)
+	}
+
+	err = s.decodeFileStore(temp.StoreType, temp.StoreConfig)
+	if err != nil {
+		return fmt.Errorf("could not decode filestore: %w", err)
+	}
+
+	return nil
+}
+
+func (s SparkConfig) MutableFields() ss.StringSet {
+	result := ss.StringSet{}
+	var executor ss.StringSet
+	var store ss.StringSet
+
+	switch s.ExecutorType {
+	case EMR:
+		executor = s.ExecutorConfig.(*EMRConfig).MutableFields()
+	case Databricks:
+		executor = s.ExecutorConfig.(*DatabricksConfig).MutableFields()
+	}
+
+	switch s.StoreType {
+	case Azure:
+		store = s.StoreConfig.(*AzureFileStoreConfig).MutableFields()
+	case S3:
+		store = s.StoreConfig.(*S3FileStoreConfig).MutableFields()
+	case GCS:
+		store = s.StoreConfig.(*GCSFileStoreConfig).MutableFields()
+	}
+
+	for field, val := range executor {
+		result["Executor."+field] = val
+	}
+
+	for field, val := range store {
+		result["Store."+field] = val
+	}
+
+	return result
+}
+
+func (a SparkConfig) DifferingFields(b SparkConfig) (ss.StringSet, error) {
+	result := ss.StringSet{}
+	var executor ss.StringSet
+	var store ss.StringSet
+	var err error
+
+	switch a.ExecutorType {
+	case EMR:
+		if b.ExecutorType != EMR {
+			err = fmt.Errorf("executor config mismatch: a = %v; b = %v", a.ExecutorType, b.ExecutorType)
+		} else {
+			executor, err = a.ExecutorConfig.(*EMRConfig).DifferingFields(*b.ExecutorConfig.(*EMRConfig))
```

Comment:
Change this to executorFields (and storeFields below). This looks like its returning an executor